File: packages/sandbox/daemon/config-store/store.ts
import type { EnrichedTenantConfig, TenantConfig } from "../types";
import { validateTenantConfig } from "../validate";
import { classify } from "./classify";
import { enrich } from "./derive";
import { deepMerge } from "./merge";
import { REJECTION_REASONS, type ApplyEvent, type ApplyResult } from "./types";

type Compute = (current: TenantConfig | null) => Partial<TenantConfig> | null;

interface QueueEntry {
  patch?: Partial<TenantConfig>;
  compute?: Compute;
  /** When true, skip subscriber notification (e.g. orchestrator-internal fills). */
  silent?: boolean;
  resolve: (r: ApplyResult) => void;
}

/**
 * Single-writer, in-memory store for tenant config.
 *
 * - All mutations go through `apply()` / `applyInternal()`. An internal FIFO
 *   worker drains pending applies one at a time, so two concurrent PUT
 *   /config requests compose deterministically (last write wins on the same
 *   field).
 * - `subscribe()` listeners run synchronously inside the worker after each
 *   applied change. Subscribers must return immediately — slow handlers
 *   stall the queue.
 * - Nothing is persisted: `.decocms/daemon.json` is read-only at boot and
 *   any further state lives only in memory until the next daemon restart.
 */
export class TenantConfigStore {
  private current: EnrichedTenantConfig | null = null;
  private readonly subscribers = new Set<(e: ApplyEvent) => void>();
  private readonly queue: QueueEntry[] = [];
  private draining = false;

  read(): EnrichedTenantConfig | null {
    return this.current;
  }

  /**
   * Bootstrap the in-memory state from a value already on disk (or seeded
   * from env). Does NOT classify, persist, or notify subscribers — purely
   * loads memory. Used once during daemon boot, before the HTTP server
   * accepts requests, so it can safely skip the apply queue.
   */
  hydrate(config: TenantConfig): void {
    this.current = enrich(config);
  }

  /**
   * Drop in-memory state. Used on orchestrator failure to reset to
   * "awaiting fresh bootstrap."
   */
  clear(): void {
    this.current = null;
  }

  apply(patch: Partial<TenantConfig>): Promise<ApplyResult> {
    return new Promise((resolve) => {
      this.queue.push({ patch, resolve });
      void this.drain();
    });
  }

  /**
   * Apply a patch computed from the post-queue state without notifying
   * subscribers. The orchestrator uses this during bootstrap to fill missing
   * fields (lockfile-detected pm/runtime, disk fallback) without re-emitting
   * pm-change/runtime-change — it already runs install+start in the same
   * pass, so a fresh transition would re-trigger them.
   *
   * `compute` runs inside the queue worker, AFTER any earlier queued applies
   * have settled. This eliminates the race where reading state out-of-band
   * and then writing it back would clobber a concurrent PUT.
   *
   * Return `null` from `compute` to skip the apply (no-op).
   */
  applyInternal(compute: Compute): Promise<ApplyResult> {
    return new Promise((resolve) => {
      this.queue.push({ compute, silent: true, resolve });
      void this.drain();
    });
  }

  subscribe(fn: (e: ApplyEvent) => void): () => void {
    this.subscribers.add(fn);
    return () => {
      this.subscribers.delete(fn);
    };
  }

  private async drain(): Promise<void> {
    if (this.draining) return;
    this.draining = true;
    try {
      while (this.queue.length > 0) {
        const entry = this.queue.shift();
        if (!entry) break;
        try {
          entry.resolve(await this.runOne(entry));
        } catch {
          entry.resolve({
            kind: "rejected",
            reason: REJECTION_REASONS.APPLY_FAILED,
          });
        }
      }
    } finally {
      this.draining = false;
    }
  }

  private async runOne(entry: QueueEntry): Promise<ApplyResult> {
    const before = this.current ? plainConfig(this.current) : null;
    const patch = entry.compute ? entry.compute(before) : entry.patch;
    if (!patch) {
      return {
        kind: "applied",
        before,
        after: before ?? {},
        transition: { kind: "no-op" },
      };
    }
    const merged = deepMerge(before, patch);

    const validation = validateTenantConfig(merged);
    if (validation.kind === "invalid") {
      return { kind: "rejected", reason: REJECTION_REASONS.INVALID };
    }

    const transition = classify(before, merged);
    if (transition.kind === "identity-conflict") {
      return {
        kind: "rejected",
        reason: REJECTION_REASONS.IMMUTABLE,
        detail: transition.field,
      };
    }

    if (transition.kind === "no-op") {
      return {
        kind: "applied",
        before,
        after: merged,
        transition,
      };
    }

    this.current = enrich(merged);

    if (!entry.silent) {
      const event: ApplyEvent = { before, after: merged, transition };
      for (const sub of this.subscribers) {
        try {
          sub(event);
        } catch {
          /* one bad subscriber does not stall the queue */
        }
      }
    }

    return {
      kind: "applied",
      before,
      after: merged,
      transition,
    };
  }
}

/** Strip in-memory derived fields when reading "before" out of state. */
function plainConfig(enriched: EnrichedTenantConfig): TenantConfig {
  return {
    git: enriched.git,
    application: enriched.application,
  };
}

File: packages/sandbox/daemon-go/internal/config/store.go
package config

import "sync"

const (
	RejectInvalid     = "invalid"
	RejectImmutable   = "immutable"
	RejectApplyFailed = "apply failed"
)

type ApplyResult struct {
	Applied    bool
	Reason     string
	Detail     string
	Before     *TenantConfig
	After      *TenantConfig
	Transition Transition
}

type ApplyEvent struct {
	Before     *TenantConfig
	After      *TenantConfig
	Transition Transition
}

type Enriched struct {
	TenantConfig
	RuntimePathPrefix string
}

func RuntimePathPrefix(runtime string) string {
	switch runtime {
	case "bun":
		return "export PATH=/opt/bun/bin:$PATH && "
	case "deno":
		return "export PATH=/opt/deno/bin:$PATH && "
	}
	return ""
}

// Store is the single-writer tenant config store: all mutations serialize
// through the apply mutex, subscribers run synchronously inside it.
type Store struct {
	mu          sync.Mutex
	current     *Enriched
	subscribers []func(ApplyEvent)
}

func NewStore() *Store {
	return &Store{}
}

func (s *Store) Read() *Enriched {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.current
}

func (s *Store) Hydrate(c *TenantConfig) {
	s.mu.Lock()
	s.current = enrich(c)
	s.mu.Unlock()
}

func (s *Store) Clear() {
	s.mu.Lock()
	s.current = nil
	s.mu.Unlock()
}

func (s *Store) Subscribe(fn func(ApplyEvent)) {
	s.mu.Lock()
	s.subscribers = append(s.subscribers, fn)
	s.mu.Unlock()
}

func (s *Store) Apply(patch *Patch) ApplyResult {
	return s.apply(func(*TenantConfig) *Patch { return patch }, false)
}

// ApplyInternal applies a patch computed from the current state without
// notifying subscribers (orchestrator default fills).
func (s *Store) ApplyInternal(compute func(current *TenantConfig) *Patch) ApplyResult {
	return s.apply(compute, true)
}

func (s *Store) apply(compute func(current *TenantConfig) *Patch, silent bool) ApplyResult {
	s.mu.Lock()

	var before *TenantConfig
	if s.current != nil {
		c := s.current.TenantConfig
		before = &c
	}
	patch := compute(before)
	if patch == nil {
		s.mu.Unlock()
		after := before
		if after == nil {
			after = &TenantConfig{}
		}
		return ApplyResult{Applied: true, Before: before, After: after, Transition: Transition{Kind: KindNoOp}}
	}
	merged := DeepMerge(before, patch)

	if reason := Validate(merged); reason != "" {
		s.mu.Unlock()
		return ApplyResult{Reason: RejectInvalid, Detail: reason}
	}

	transition := Classify(before, merged)
	if transition.Kind == KindIdentityConflict {
		s.mu.Unlock()
		return ApplyResult{Reason: RejectImmutable, Detail: transition.Field}
	}

	if transition.Kind == KindNoOp {
		// Still persist. "No-op" classifies the SIDE EFFECT (no re-clone, no
		// restart, no subscriber wake-up), not the write: a patch can carry a
		// field Classify has no arm for — submodule credentials are the live
		// case — and returning `merged` to the caller while dropping it from the
		// store answers 200 with a receipt for a write that never happened.
		// Guarded on non-nil so an inert patch can't claim an unclaimed daemon.
		if s.current != nil {
			s.current = enrich(merged)
		}
		s.mu.Unlock()
		return ApplyResult{Applied: true, Before: before, After: merged, Transition: transition}
	}

	s.current = enrich(merged)
	subs := append([]func(ApplyEvent){}, s.subscribers...)
	s.mu.Unlock()

	if !silent {
		event := ApplyEvent{Before: before, After: merged, Transition: transition}
		for _, fn := range subs {
			func() {
				defer func() { recover() }()
				fn(event)
			}()
		}
	}

	return ApplyResult{Applied: true, Before: before, After: merged, Transition: transition}
}

func enrich(c *TenantConfig) *Enriched {
	return &Enriched{
		TenantConfig:      *c,
		RuntimePathPrefix: RuntimePathPrefix(c.Runtime()),
	}
}
